spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -327,19 +340,18 @@ public Boolean execute(RedisOperations ops) throws DataAccessException {
 
 					ops.watch(key); //monitor key
 
-					ops.multi(); //transaction start
+					if (ops.opsForValue().get(key) != null) {
+						ops.unwatch(); //key already exists, stop monitoring
+						return false;
+					}
 
-					//can't rely on operations result inside transaction, execution is delayed till `exec()`
-					ops.opsForValue().setIfAbsent(key, RedisLock.this);
+					ops.multi(); //transaction start
 
-					//set expire on key if exists
-					ops.expire(key, RedisLockRegistry.this.expireAfter, TimeUnit.MILLISECONDS);
+					//set the value and expire
+					ops.opsForValue().set(key, RedisLock.this, RedisLockRegistry.this.expireAfter, TimeUnit.MILLISECONDS);
 
 					//exec will contain all operations result or null - if execution has been aborted due to 'watch'
-					List result = ops.exec();
-
-					//check 'setIfAbsent' result (first in list)
-					return (result != null) && (!result.isEmpty()) && (Boolean.TRUE.equals(result.get(0)));
+					return ops.exec() != null;
```

Comment:
Hello Artem!
Yes I'm fixing own mistake - improper expire update. `setIfAbsent` was under same `watch` as new statement does, and yes - transactional model works. All modifications within `watch` and `exec` will cause `exec` to fail (return null in our case) and monitoring of the key starts right from `watch`. Easy to check in debug (I checked) - just put some `set` between `watch` and `multi`.